### PR TITLE
VP-4132: Cart XAPI Security

### PR DESCRIPTION
**Pull request by AliveMen**
### Problem
Cart search results not filtered by user that belongs to.

### Solution
Add security
### Make sure these boxes are checked:
- [x] Check all the changes in github PR - files count (non of them are redundant, have meaningful changes, all are added), if target branch is correct
- [x] Check methods and variable namings - it should be self descriptive, no typos
- [x] Check you did not introduce breaking changes in API and public models/services.
- [x] Respect extensibility - https://community.virtocommerce.com/t/extensibility-basics-the-domain-model-and-persistence-layer-extension/141
- [x] Follow [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) and [SOLID](https://en.wikipedia.org/wiki/SOLID) principles
- [x] For unit tests - follow Microsoft best practices: https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices
- [x] Consolidate solution dependencies in case you are using newer version, update VC module dependencies in module.manifest. Do not upgrade 3rd party packages that are shipped with the platform with the version newer than in the platform.
- [x] Check code style conventions - https://github.com/VirtoCommerce/styleguide/blob/master/csharp.md
- [x] Check PR have a concise and descriptive title, follow [git commit message rules](https://github.com/VirtoCommerce/styleguide/blob/master/gitcommits.md)

### CODE REVIEW

File: src/XPurchase/VirtoCommerce.XPurchase/Extensions/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -14,6 +18,7 @@ public static class ServiceCollectionExtensions
     {
         public static IServiceCollection AddXPurchase(this IServiceCollection services, IGraphQLBuilder graphQlbuilder)
         {
+            services.AddSingleton<Func<SignInManager<ApplicationUser>>>(provider => () => provider.GetService<SignInManager<ApplicationUser>>());
```
Comment (yecli) [inline src/XPurchase/VirtoCommerce.XPurchase/Extensions/ServiceCollectionExtensions.cs:21]: Please remove, as now we registered this in the platform. I've already updated dependency

File: src/XPurchase/VirtoCommerce.XPurchase/Schemas/PurchaseSchema.cs [hunk 2]
```diff
@@ -643,5 +654,37 @@ private async Task<object> ResolveConnectionAsync(IMediator mediator, IResolveCo
...
+
+            if (user == null)
+            {
+                user = new ApplicationUser()
+                {
+                    Id = userId,
+                    UserName = "Anonymous",
+                };
+                
+                await signInManager.ClaimsFactory.CreateAsync(user);
```
Comment (yecli) [inline src/XPurchase/VirtoCommerce.XPurchase/Schemas/PurchaseSchema.cs:678]: Here we already got ClaimPrincilal, no need to signInManager.CreateUserPrincipalAsync later.

File: src/XPurchase/VirtoCommerce.XPurchase/Extensions/ServiceCollectionExtensions.cs (hunk 1 above)
Comment (AliveMen) [reply src/XPurchase/VirtoCommerce.XPurchase/Extensions/ServiceCollectionExtensions.cs]: Removed

File: src/XPurchase/VirtoCommerce.XPurchase/Schemas/PurchaseSchema.cs (hunk 2 above)
Comment (AliveMen) [reply src/XPurchase/VirtoCommerce.XPurchase/Schemas/PurchaseSchema.cs]: Fixed

File: src/XPurchase/VirtoCommerce.XPurchase/Schemas/PurchaseSchema.cs [hunk 3]
```diff
@@ -643,5 +654,30 @@ private async Task<object> ResolveConnectionAsync(IMediator mediator, IResolveCo
...
+            var user = await signInManager.UserManager.FindByIdAsync(userId) ?? new ApplicationUser
+            {
+                Id = userId,
+                UserName = "Anonymous",
+            };
+
+            var userPrincipal = await signInManager.ClaimsFactory.CreateAsync(user);
+            var authorizationResult = await _authorizationService.AuthorizeAsync(userPrincipal, resource, new CanAccessCartAuthorizationRequirement());
+
+            if (!authorizationResult.Succeeded)
```
Comment (tatarincev) [inline src/XPurchase/VirtoCommerce.XPurchase/Schemas/PurchaseSchema.cs:677]: Please check how to returns 401.
https://github.com/VirtoCommerce/vc-module-experience-api/blob/dev/src/XProfile/VirtoCommerce.XProfile/Schemas/OrganizationType.cs#L21

Comment (yecli) [reply src/XPurchase/VirtoCommerce.XPurchase/Schemas/PurchaseSchema.cs:677]: Returning 401 is possible only by overring GraphQL Http server, parsing error messages and integrating our imperative auth checks into their auth system: https://github.com/graphql-dotnet/authorization/issues/1
I think it does not worth it, let's leave it as is.

**Pull request merged**